src/main/java/uk/gov/pay/directdebit/payments/dao/DirectDebitEventDao.java

```diff
@@ -1,26 +1,114 @@
+    private static final String QUERY = "SELECT e.id, e.mandate_id, e.transaction_id, e.event_type, e.event, e.event_date, e.version, e.external_id, m.external_id AS mandate_external_id, t.external_id AS transaction_external_id " +
+            "FROM events e LEFT OUTER JOIN mandates m ON (e.mandate_id = m.id) " +
+            "LEFT OUTER JOIN transactions t ON (e.transaction_id = t.id) " +
+            ":searchFields ORDER BY e.id DESC LIMIT :limit OFFSET :offset";
+    private static final String COUNT_QUERY = "SELECT count(*) " +
+            "FROM events e LEFT OUTER JOIN mandates m ON (e.mandate_id = m.id) " +
+            "LEFT OUTER JOIN transactions t ON (e.transaction_id = t.id) " +
+            ":searchFields";
+    
+    private final Jdbi jdbi;
 
-    @SqlQuery("SELECT * FROM events e WHERE e.mandate_id = :mandateId and e.event_type = :eventType and e.event = :event")
-    Optional<DirectDebitEvent> findByMandateIdAndEvent(@Bind("mandateId") Long mandateId, @Bind("eventType") Type eventType, @Bind("event") SupportedEvent event);
+    @Inject
+    public DirectDebitEventDao(Jdbi jdbi) {
+        this.jdbi = jdbi;
+    }
+    
+    public List<DirectDebitEvent> findEvents(DirectDebitEventSearchParams searchParams) {
+        QueryStringAndQueryMap queryStringAndQueryMap = generateQuery(searchParams);
+        String limit = isNull(searchParams.getPageSize()) ? "NULL" : searchParams.getPageSize().toString();
+        String offset = searchParams.getPage() == null ? "0" : Integer.toString((searchParams.getPage() - 1) * searchParams.getPageSize());
+        return jdbi.withHandle(handle -> {
+            Query query = handle.createQuery(QUERY
+                    .replace(":searchFields", queryStringAndQueryMap.queryString)
+                    .replace(":limit", limit)
+                    .replace(":offset", offset));
+            queryStringAndQueryMap.queryMap.forEach(query::bind);
+            return query.map(new EventMapper()).list();
+        });
+    }
+
+    private QueryStringAndQueryMap generateQuery(DirectDebitEventSearchParams searchParams) {
+        List<String> searchStrings = Lists.newArrayList();
+        Map<String, Object> queryMap = Maps.newHashMap();
+        if (nonNull(searchParams.getMandateExternalId())) {
+            searchStrings.add("m.external_id = :mandate_id");
+            queryMap.put("mandate_id", searchParams.getMandateExternalId());
+        }
+        if (nonNull(searchParams.getTransactionExternalId())) {
+            searchStrings.add("t.external_id = :transaction_id");
+            queryMap.put("transaction_id", searchParams.getTransactionExternalId());
+        }
+        if (nonNull(searchParams.getBeforeDate())) {
+            searchStrings.add("e.event_date < :before_date");
+            queryMap.put("before_date", searchParams.getBeforeDate());
+        }
+        if (nonNull(searchParams.getAfterDate())) {
+            searchStrings.add("e.event_date > :after_date");
```

Comment:
I think `>=` might be more consistent with what we do for other date-based searches